### PR TITLE
Redirect (console) output (expose as signal) 

### PR DESCRIPTION
**Pull request by kuhnchris**
This is the master/4.x version of #54525.
Tested on Godot master 27th Jan '22 13:30pm.

Example test project:

[gd4-example.zip](https://github.com/godotengine/godot/files/7950420/gd4-example.zip)

Same things apply as for #54525 - I'm a new contributor so please let me know if we want to merge it into a different existing Singleton.

### CODE REVIEW

**Comment by slapin** [comment]
Thanks a lot for your work.
Could you please make meaningful subject line which would describe PR in a few words as now it is impossible to guess what this PR is about without entering it and reading the long description. Thanks!

**Comment by kuhnchris** [comment]
Thanks, I knew I forgot something when I opened the PR...

**Comment by Zireael07** [comment]
For reference, the commit hash as of your testing is a743bce (in case some future changes break this PR, this would be a "known good" commit)

**Comment by qarmin** [comment]
There is memory leak found and can be fixed by adding memdelete 

https://github.com/godotengine/godot/pull/49123/files#diff-18b7c19417a461e4793bb2157831659d96ede991391e10ee5bc51f19453b7685R1409

**Comment by kuhnchris** [comment]
> There is memory leak found and can be fixed by adding memdelete
> 
> https://github.com/godotengine/godot/pull/49123/files#diff-18b7c19417a461e4793bb2157831659d96ede991391e10ee5bc51f19453b7685R1409

Thanks, I removed it similar to the IP object in unregister_core_types.

-Chris

File: doc/classes/Redirect.xml [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+	<brief_description>
+	</brief_description>
+	<description>
+	</description>
+	<tutorials>
+	</tutorials>
+	<signals>
+		<signal name="print_line">
+			<argument index="0" name="scene_root" type="Node" />
+			<description>
```
Comment (mhilbrunner) [inline doc/classes/Redirect.xml:12]: Please will in the docs/description for the new signal here in `<description></description>` :)

**Comment by mhilbrunner** [comment]
Looks good (and useful) to me, although I wonder about the performance.

File: doc/classes/Redirect.xml (hunk 1 above)
Comment (kuhnchris) [reply doc/classes/Redirect.xml:12]: yessir, on it, will do it in a sec.

Comment (kuhnchris) [reply doc/classes/Redirect.xml:15]: Nevermind, after recompiling I have the text in the editor :-)

**Comment by akien-mga** [comment]
This looks like something that should go through our existing Logger API: https://github.com/godotengine/godot/blob/master/core/io/logger.h

It's designed specifically to allow overriding the way logging works, so you shouldn't need to add a new singleton for this.

@reduz also pointed out that using a signal will likely cause issues as it's not thread safe.

**Comment by akien-mga** [comment]
This needs further work as pointed out in my previous comment but didn't really get updated after the initial draft implementation, so let's close it for now as it's not in a state which is ready to merge.

It still seems like an interesting feature but the implementation needs more research to be done properly. I would suggest that an interested contributor opens a proposal on https://github.com/godotengine/godot-proposals to describe how this feature is meant to be used, behave and be implemented (which should have been done before this PR normally), so we can keep track of it and of eventual future implementation attempts.

**Pull request closed without merging**